File: bindings/java/src/main/java/com/aster/interceptors/RateLimitInterceptor.java
package com.aster.interceptors;

import java.util.concurrent.ConcurrentHashMap;

/**
 * Token bucket rate limiter interceptor.
 *
 * <p>Limits request rate per service, per method, or per peer. Rejects requests that exceed the
 * limit with {@link StatusCode#RESOURCE_EXHAUSTED}.
 */
public final class RateLimitInterceptor implements Interceptor {

  private final double rate;
  private final double burst;
  private final String per;
  private final ConcurrentHashMap<String, TokenBucket> buckets = new ConcurrentHashMap<>();
  private final TokenBucket globalBucket;

  /**
   * Creates a rate limiter.
   *
   * @param rate maximum requests per second
   * @param burst maximum burst size (defaults to rate if &lt;= 0)
   * @param per granularity: "global", "service", "method", or "peer"
   */
  public RateLimitInterceptor(double rate, double burst, String per) {
    this.rate = rate;
    this.burst = burst > 0 ? burst : rate;
    this.per = per != null ? per : "global";
    this.globalBucket = new TokenBucket(rate, this.burst);
  }

  /** Creates a global rate limiter with the given rate and burst equal to rate. */
  public RateLimitInterceptor(double rate) {
    this(rate, rate, "global");
  }

  @Override
  public Object onRequest(CallContext ctx, Object request) {
    TokenBucket bucket = getBucket(ctx);
    if (!bucket.tryAcquire()) {
      throw new RpcError(
          StatusCode.RESOURCE_EXHAUSTED, "Rate limit exceeded (" + rate + "/s per " + per + ")");
    }
    return request;
  }

  private TokenBucket getBucket(CallContext ctx) {
    if ("global".equals(per)) {
      return globalBucket;
    }

    String key;
    switch (per) {
      case "service" -> key = ctx.service();
      case "method" -> key = ctx.service() + "." + ctx.method();
      case "peer" -> key = ctx.peer() != null ? ctx.peer() : "unknown";
      default -> {
        return globalBucket;
      }
    }

    return buckets.computeIfAbsent(key, k -> new TokenBucket(rate, burst));
  }

  /** Simple token bucket rate limiter (thread-safe). */
  private static final class TokenBucket {
    private final double rate;
    private final double capacity;
    private double tokens;
    private long lastRefillNanos;

    TokenBucket(double rate, double capacity) {
      this.rate = rate;
      this.capacity = capacity;
      this.tokens = capacity;
      this.lastRefillNanos = System.nanoTime();
    }

    synchronized boolean tryAcquire() {
      long now = System.nanoTime();
      double elapsed = (now - lastRefillNanos) / 1_000_000_000.0;
      tokens = Math.min(capacity, tokens + elapsed * rate);
      lastRefillNanos = now;

      if (tokens >= 1.0) {
        tokens -= 1.0;
        return true;
      }
      return false;
    }
  }
}

File: bindings/dotnet/src/Aster/Interceptors/RateLimitInterceptor.cs
namespace Aster.Interceptors;

/// <summary>
/// Token bucket rate limiter interceptor.
/// Limits request rate per service, per method, per peer, or globally.
/// Rejects requests that exceed the limit with ResourceExhausted status.
/// </summary>
public class RateLimitInterceptor : IInterceptor
{
    private readonly double _rate;
    private readonly double? _burst;
    private readonly string _per;
    private readonly Dictionary<string, TokenBucket> _buckets = new();
    private readonly TokenBucket _globalBucket;

    /// <param name="rate">Maximum requests per second.</param>
    /// <param name="burst">Maximum burst size (defaults to rate).</param>
    /// <param name="per">Granularity: "global", "service", "method", or "peer".</param>
    public RateLimitInterceptor(double rate = 100.0, double? burst = null, string per = "global")
    {
        _rate = rate;
        _burst = burst;
        _per = per;
        _globalBucket = new TokenBucket(rate, burst ?? rate);
    }

    public object OnRequest(CallContext ctx, object request)
    {
        var bucket = GetBucket(ctx);
        if (!bucket.TryAcquire())
        {
            throw new RpcError(
                StatusCode.ResourceExhausted,
                $"Rate limit exceeded ({_rate}/s per {_per})");
        }
        return request;
    }

    private TokenBucket GetBucket(CallContext ctx)
    {
        if (_per == "global")
            return _globalBucket;

        var key = _per switch
        {
            "service" => ctx.Service,
            "method" => $"{ctx.Service}.{ctx.Method}",
            "peer" => ctx.Peer ?? "unknown",
            _ => "global"
        };

        if (key == "global")
            return _globalBucket;

        if (!_buckets.TryGetValue(key, out var bucket))
        {
            bucket = new TokenBucket(_rate, _burst ?? _rate);
            _buckets[key] = bucket;
        }
        return bucket;
    }

    /// <summary>
    /// Simple token bucket rate limiter.
    /// </summary>
    private sealed class TokenBucket
    {
        private readonly double _rate;
        private readonly double _capacity;
        private double _tokens;
        private long _lastRefillTicks;

        public TokenBucket(double rate, double capacity)
        {
            _rate = rate;
            _capacity = capacity;
            _tokens = capacity;
            _lastRefillTicks = Environment.TickCount64;
        }

        public bool TryAcquire()
        {
            long now = Environment.TickCount64;
            double elapsed = (now - _lastRefillTicks) / 1000.0;
            _tokens = Math.Min(_capacity, _tokens + elapsed * _rate);
            _lastRefillTicks = now;

            if (_tokens >= 1.0)
            {
                _tokens -= 1.0;
                return true;
            }
            return false;
        }
    }
}
